beacon_node/http_api/src/lib.rs

```diff
@@ -2356,6 +2356,41 @@ pub fn serve<T: BeaconChainTypes>(
             })
         });
 
+    // GET lighthouse/seen_validators?ids,epochs
+    let get_lighthouse_seen_validators = warp::path("lighthouse")
+        .and(warp::path("seen_validators"))
+        .and(warp::path::end())
+        .and(warp::query::<api_types::SeenValidatorQuery>())
+        .and(chain_filter.clone())
+        .and_then(
+            |query: api_types::SeenValidatorQuery, chain: Arc<BeaconChain<T>>| {
+                blocking_json_task(move || {
+                    let indices: Vec<usize> = chain
+                        .head_beacon_state()
```

Comment:
We could avoid cloning a `BeaconState` here and use the `validator_pubkey_cache`:

https://github.com/sigp/lighthouse/blob/46920a84e857d6c086adf4b42810ac03de94ed3e/beacon_node/beacon_chain/src/beacon_chain.rs#L728-L747

Reply:
Made a bunch of updates here, and ready for another round of review. To address some of the questions posed earlier @paulhauner 

> 1. We will need to consider how this performs prior to genesis. I think that if we start the VC prior to genesis we want to skip DGP (doppelganger protection) completely. Otherwise, we'll end up with no nodes producing blocks for the first couple of epochs.

I've made it so if doppelganger detection is enabled for a validator at or before the genesis epoch, nothing happens (validator runs normally and doesn't check for doppelgangers)

> 2. I think we might want to turn on DGP protection by default, since it's the safest route. Users can opt into fast and risky if they want.

I've set it to run on default and added an opt out flag. I do think that there's a risk we will have a lot of people upset about missing ~3 attestations per validator on restart if we aren't clear enough about the change though. Also, do you think this presents any risk around lighthouse releases (maybe an emergency hotfix or something)? if some large portion of the network updates at once?

> 3. I wonder if this will falsely detect a doppelganger if we reboot in the middle of the epoch. E.g., if the validator attested in the first slot, we reboot in the second slot and then we check for DGs in the third slot and then get a hit because we find our validator in the same epoch.

I've made it so whenever the validator starts, it detects attestations in (current epoch + 1) and (current epoch + 2), until the end of (current epoch + 2). So I don't think we will detect anything on reboots, unless there is an edge case in there around clock drift?

> 4. What are your thoughts on writing DGP as a separate routine which we run before starting the block/attestation/duties services? It might turn out a little simpler to implement and perhaps less likely for us to forget to gate something behind it in the future. I haven't thought deeply about this, just raising it :)

Separated into a separate service :)